keps/sig-node/20200619-provisioning-resources-with-numa-topology.md

```diff
@@ -0,0 +1,240 @@
+---
+title: Provisioning the Node Resources With NUMA Topology Information
+authors:
+  - "@AlexeyPerevalov"
+  - "@swatisehgal"
+owning-sig: sig-node
+participating-sigs:
+  - sig-node
+  - sig-scheduling
+reviewers:
+  - "@dchen1107"
+  - "@derekwaynecarr"
+approvers:
+  - "@dchen1107"
+  - "@derekwaynecarr"
+creation-date: 2020-06-19
+last-updated: 2020-06-19
+status: implementable
+see-also:
+  - "/keps/sig-scheduling/20200612-deducted-topology-manager.md"
+---
+
+# Exposure Node Resources With NUMA Topology Information
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Design based on CRI](#cri)
+  - [Design based on podresources interface of the kubelet](#podresources)
+  - [API](#api)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+  - [Annotation approach](#annotation-approach)
+  - [NUMA specification in ResourceName](#numa-in-resourcename)
+<!-- /toc -->
+
+## Summary
+
+Cluster which contains several nodes with NUMA topology and
+enabled TopologyManager feature gate is not rare now. In such cluster
+could be a situation when TopologyManager's admit handler on the kubelet
+side refuses pod launching since pod's resources request doesn't sutisfy
+selected TopologyManager policy, in this case pod should be rescheduled.
+And so it can get stuck in the rescheduling cycle.
```

Comment:
this solution assumes the desired outcome is for a pod to remain pending until a viable node is found.  another viable solution is for the scheduler to be aware of historical scheduling decisions of similar pods that reached an unexpected terminal state prior to running.

is this solution optimizing for scheduling time?  how quickly do we think a scheduler can make a decision in this situation, especially where secondary affinity across devices is not obvious?  for example, the hints that were added to the device plugin that lets the kubelet understand a preferred set of co-located devices based on device plugin itself.

/cc @klueska

Reply:
AFAIK, If the pod is determined to be unschedulable or if there is an internal error it is returned to the scheduling queue and retried. While the scheduling of the pod is being retried, it appears to the end user in pending state. @ahg-g @Huang-Wei Can you please confirm the above, also regarding Derek's question is there currently a way to be aware of historical decisions/failures of a pod?

Regarding, optimizing the solution for scheduling time, this daemon creates CRDs based on resource information gathered from CRD/PodResource API. [#1884](https://github.com/kubernetes/enhancements/pull/1884/files) aims to allow gathering device information along with numa ids from PodResource API.